L1Trigger/L1TTrackMatch/plugins/DisplacedVertexProducer.cc

```diff
@@ -0,0 +1,221 @@
+  if((R>=(R1-R2)) && (R<=(R1+R2))){
+    return (0);
+  }
+  else if(R==0){
+    return (-99999.0);
+  }
+  else{
+    return(R-R1-R2);
+  }
+}
+
+Int_t calcVertex(Track_Parameters a, Track_Parameters b, Double_t &x_vtx, Double_t &y_vtx, Double_t &z_vtx){
+  float x1 = a.x0; //   Centers of the circles
+  float y1 = a.y0; //
+  float x2 = b.x0; //
+  float y2 = b.y0; //
+  float R1 = a.rho;   // Radii of the circles
+  float R2 = b.rho;
+  float R = dist(x1,y1,x2,y2); // Distance between centers
+  if(R==0) return -1;
+  float co1 = (pow(R1,2)-pow(R2,2))/(2*pow(R,2));
+  float radicand = (2/pow(R,2))*(pow(R1,2)+pow(R2,2))-(pow(pow(R1,2)-pow(R2,2),2)/pow(R,4))-1;
+  float co2 = 0;
+  if(radicand>0) co2 = 0.5*TMath::Sqrt(radicand);
+  float ix1_x = 0.5*(x1+x2)+co1*(x2-x1)+co2*(y2-y1);
+  float ix2_x = 0.5*(x1+x2)+co1*(x2-x1)-co2*(y2-y1);
+  float ix1_y = 0.5*(y1+y2)+co1*(y2-y1)+co2*(x1-x2);
+  float ix2_y = 0.5*(y1+y2)+co1*(y2-y1)-co2*(x1-x2);
+  float ix1_z1 = a.z(ix1_x,ix1_y);
+  float ix1_z2 = b.z(ix1_x,ix1_y);
+  float ix1_delz = fabs(ix1_z1-ix1_z2);
+  float ix2_z1 = a.z(ix2_x,ix2_y);
+  float ix2_z2 = b.z(ix2_x,ix2_y);
+  float ix2_delz = fabs(ix2_z1-ix2_z2);
+  float trk1_POCA[2] = {a.d0*sin(a.phi),-1*a.d0*cos(a.phi)};
+  float trk2_POCA[2] = {b.d0*sin(b.phi),-1*b.d0*cos(b.phi)};
+  float trk1_ix1_delxy[2] = {ix1_x-trk1_POCA[0],ix1_y-trk1_POCA[1]};
+  float trk1_ix2_delxy[2] = {ix2_x-trk1_POCA[0],ix2_y-trk1_POCA[1]};
+  float trk2_ix1_delxy[2] = {ix1_x-trk2_POCA[0],ix1_y-trk2_POCA[1]};
+  float trk2_ix2_delxy[2] = {ix2_x-trk2_POCA[0],ix2_y-trk2_POCA[1]};
+  float trk1_traj[2] = {cos(a.phi),sin(a.phi)};
+  float trk2_traj[2] = {cos(b.phi),sin(b.phi)};
+  bool trk1_ix1_inTraj = ((trk1_ix1_delxy[0]*trk1_traj[0]+trk1_ix1_delxy[1]*trk1_traj[1])>0) ? true : false;
+  bool trk1_ix2_inTraj = ((trk1_ix2_delxy[0]*trk1_traj[0]+trk1_ix2_delxy[1]*trk1_traj[1])>0) ? true : false;
+  bool trk2_ix1_inTraj = ((trk2_ix1_delxy[0]*trk2_traj[0]+trk2_ix1_delxy[1]*trk2_traj[1])>0) ? true : false;
+  bool trk2_ix2_inTraj = ((trk2_ix2_delxy[0]*trk2_traj[0]+trk2_ix2_delxy[1]*trk2_traj[1])>0) ? true : false;
+  if(trk1_ix1_inTraj&&trk2_ix1_inTraj&&trk1_ix2_inTraj&&trk2_ix2_inTraj){
+    if(ix1_delz<ix2_delz){
+      x_vtx = ix1_x;
+      y_vtx = ix1_y;
+      z_vtx = (ix1_z1+ix1_z2)/2;
+      return 0;
+    }
+    else{
+      x_vtx = ix2_x;
+      y_vtx = ix2_y;
+      z_vtx = (ix2_z1+ix2_z2)/2;
+      return 0;
+    }
+  }
+  if(trk1_ix1_inTraj&&trk2_ix1_inTraj){
+    x_vtx = ix1_x;
+    y_vtx = ix1_y;
+    z_vtx = (ix1_z1+ix1_z2)/2;
+    return 1;
+  }
+  if(trk1_ix2_inTraj&&trk2_ix2_inTraj){
+    x_vtx = ix2_x;
+    y_vtx = ix2_y;
+    z_vtx = (ix2_z1+ix2_z2)/2;
+    return 2;
+  }
+  else{
+    if(ix1_delz<ix2_delz){
+      x_vtx = ix1_x;
+      y_vtx = ix1_y;
+      z_vtx = (ix1_z1+ix1_z2)/2;
+      return 3;
+    }
+    else{
+      x_vtx = ix2_x;
+      y_vtx = ix2_y;
+      z_vtx = (ix2_z1+ix2_z2)/2;
+      return 3;
+    }
+  }
+  return 4;
```

Comment:
The final else here on line 89 only considers the previous if on line 83 and not the if's on 77 and 63. Are you sure that's what you want to do? In general there are also a lot of redundant setting lines for similar conditions, I think some of this can be cleaned up.

Reply:
Can comments also be included that maybe describe what returning each of these numbers mean? Just for user interpretability